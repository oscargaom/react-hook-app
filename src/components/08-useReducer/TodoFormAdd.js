import React from 'react';
import { useForm } from '../../hooks/useForm';

export const TodoFormAdd = ({handleAddNewTodo}) => {

    const [{ description }, handleInputChange, reset] = useForm({
        description: ""
    });
    
    // console.log(description);

    const handleSubmit = (e) => {
        e.preventDefault();

        if (description.trim().length < 1) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddNewTodo(newTodo);

        reset();
    };

    return (
        <>
            <h4> Agregar TODO </h4>
            <hr />

            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Agregar nueva tarea ..."
                    autoComplete="off"
                    value={description}
                    onChange={handleInputChange}
                />

                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                >
                    Agregar
                </button>
            </form>
        </>
    )
}
