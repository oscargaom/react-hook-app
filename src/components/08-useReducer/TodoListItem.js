import React from 'react'

export const TodoListItem = ({todo, index, handleToggle, handleRemoveOnClick}) => {

    const {id, done, desc} = todo;

    return (
        <li
            key={id}
            className="list-group-item"
        >
            <p 
                className={`${done && "complete"}`}
                onClick={() => handleToggle(id)}
            >
                {index + 1}. {desc}
            </p>

            <button
                name={id}
                className="btn-danger"
                onClick={() => { handleRemoveOnClick(id) }}
            >
                Eliminar
            </button>
        </li>
    )
}
