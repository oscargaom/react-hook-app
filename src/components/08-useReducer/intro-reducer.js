
// Se crea un arreglo con un objeto.
const initialState = [{
    id: 1,
    todo: "Solfeo",
    done: false
}];

const reducer = (state = initialState, action) => {

    if (action?.type === "agregar") {
        return [...state, action.payload];
    }

    return state;
};

const todos = reducer();

const newTodo = {
    id: 2,
    todo: "Practiar por 15 minutos",
    done: false
};

const agregarTodoAction = {
    type: "agregar",
    payload: newTodo
};

const newTodos = reducer(todos, agregarTodoAction);

console.log(todos);
console.log(newTodos);
