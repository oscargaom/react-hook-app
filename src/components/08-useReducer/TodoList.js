import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({todos, handleToggle, handleRemoveOnlick}) => {
    return (
        <ul className="list-group list-group-flush">
        {
            todos.map((todo, idx) =>
                <TodoListItem 
                    key={todo.id}
                    todo={todo} 
                    index={idx} 
                    handleToggle={handleToggle} 
                    handleRemoveOnlick={handleRemoveOnlick}
                />
            )
        }
    </ul>
    )
}
