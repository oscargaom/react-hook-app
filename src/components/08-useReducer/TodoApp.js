import React, { useReducer, useEffect } from 'react'
import "../../styles/styles.css"
import { todoReducer } from './todoReducer';
import { TodoList } from './TodoList';
import { TodoFormAdd } from './TodoFormAdd';

// const initialState = [{
//     id: new Date().getTime(),
//     desc: "Solfeo",
//     done: false
// }];

const init = () => {
    // return [{
    //     id: new Date().getTime(),
    //     desc: "Solfeo",
    //     done: false
    // }]
    return JSON.parse(localStorage.getItem("todos")) || [];
};


export const TodoApp = () => {

    // const [todos, dispatch] = useReducer(todoReducer, initialState);
    const [todos, dispatch] = useReducer(todoReducer, [], init);

    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
        // console.log('entrando a useEffect');
        // console.log(todos);
    }, [todos]);

    const handleRemoveOnClick = (todoId) => {

        // const todosJson = JSON.parse(localStorage.getItem("todos"))
        // console.log(todosJson);
        // const { target: { name } } = e;
        // console.log(name);

        const action = {
            type: "remove",
            payload: todoId
        };

        dispatch(action);
    };

    const handleToggle = (todoId) => {
        console.log(todoId);
        dispatch({
            type: "toggle",
            payload: todoId
        });
    };

    const handleAddNewTodo = (todo) => {
        
        dispatch({
            type: "add",
            payload: todo
        });
    };

    return (
        <div>
            <h1>Todo App ({todos.length}) </h1>
            <hr />

            {/* <div className="row">
                <div className="col-7">
                    <ul className="list-group list-group-flush">
                        {
                            todos.map((todo, idx) =>
                                (<li
                                    key={todo.id}
                                    className="list-group-item"
                                >
                                    <p 
                                        className={`${todo.done && "complete"}`}
                                        onClick={() => handleToggle(todo.id)}
                                    >
                                        {idx + 1}. {todo.desc}
                                    </p>

                                    <button
                                        name={todo.id}
                                        className="btn-danger"
                                        onClick={() => { handleRemoveOnClick(todo.id) }}
                                    >
                                        Eliminar
                                    </button>
                                </li>)
                            )
                        }
                    </ul>
                </div>
                <div className="col-5">
                    <h4> Agregar TODO </h4>
                    <hr />

                    <form onSubmit={handleSubmit}>
                        <input
                            type="text"
                            name="description"
                            className="form-control"
                            placeholder="Agregar nueva tarea ..."
                            autoComplete="off"
                            value={description}
                            onChange={handleInputChange}
                        />

                        <button
                            type="submit"
                            className="btn btn-outline-primary mt-1 btn-block"
                        >
                            Agregar
                        </button>
                    </form>
                </div>
            </div> */}
            <div className="row">
                <div className="col-7">
                    <TodoList
                        todos={todos}
                        handleToggle={handleToggle}
                        handleRemoveOnClick={handleRemoveOnClick}
                    />
                </div>
                <div className="col-5">
                    <TodoFormAdd 
                        handleAddNewTodo={handleAddNewTodo}
                    />
                </div>
            </div>
        </div>
    )
}
