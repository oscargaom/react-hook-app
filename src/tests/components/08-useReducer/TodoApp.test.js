import React from 'react';
import '@testing-library/jest-dom';
import { shallow, mount } from 'enzyme';
import { TodoApp } from '../../../components/08-useReducer/TodoApp';
import { demoTodos, newDemoTodoItem } from '../../fixtures/demoTodos';
import { act } from '@testing-library/react';


describe('Pruebas en el componente <TodoApp /> ', () => {
    
    const wrapper = shallow(<TodoApp />);
    
    // Vamos a simular el llamado a localStorage.setItem()
    Storage.prototype.setItem = jest.fn(() => {});
    
    test('debe mostrarse correctamente el componente', () => {
        
        expect(wrapper).toMatchSnapshot();
    });


    test('debe agregar un nuevo TODO', () => {

        /*  Utilizamos mount porque carga o renderiza todos los componentes que tiene 
            internamente TodoApp [TodoList, TodoFormAdd] de este modo podemos ver su 
            contenido en formato jsx de lo contrario solo vemos las etiquetas del 
            componente, pero cuando no es el caso utilizamos shallow.
            shallow solo renderiza el primer componente dentro del high order component,
            es decir, solo renderiza el primer componente que ser√≠a el componente padre.
        */
        const mountWrapper = mount(<TodoApp />);
        
        const handleAddNewTodo = mountWrapper.find('TodoFormAdd').prop('handleAddNewTodo');
        
        act (() => {
            handleAddNewTodo(demoTodos[0]);
        });

        act (() => {
            handleAddNewTodo(demoTodos[1]);
        });

        act (() => {
            handleAddNewTodo(demoTodos[2]);
        });

        // Se debe contar la primera vez que se carga el componente 
        const countUseEffectLoadFirstTime = 1;

        expect(mountWrapper.find('h1').text().trim()).toBe(`Todo App (${ demoTodos.length })`);
        expect(localStorage.setItem).toHaveBeenCalledTimes(demoTodos.length + countUseEffectLoadFirstTime);
        
    });
    
    test('debe eliminar un elemento TODO ', () => {
        const handleAddNewTodo = wrapper.find('TodoFormAdd').props().handleAddNewTodo;

        handleAddNewTodo(newDemoTodoItem);

        const handleRemoveOnClick = wrapper.find('TodoList').props().handleRemoveOnClick;

        handleRemoveOnClick(newDemoTodoItem.id)

        expect(wrapper.find('h1').text().trim()).toBe('Todo App (0)');
    });
    
    
})

