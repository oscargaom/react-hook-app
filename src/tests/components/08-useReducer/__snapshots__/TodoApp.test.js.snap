// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <TodoApp />  debe mostrarse correctamente el componente 1`] = `
<div>
  <h1>
    Todo App (
    0
    ) 
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        handleRemoveOnClick={[Function]}
        handleToggle={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoFormAdd
        handleAddNewTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
